### ISSUE TITLE
[TW-1761] Corrupted pending.data file

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#1785]
_Adam O on 2016-01-26T18:35:02Z says:_ 

 Any taskwarrior command reports 'Unrecognized Taskwarrior file format. in /Users/xxxxxxxx/.task/pending.data at line xxx'. Upon inspection, pending.data is corrupt.

By comparing pending.data to a version from the day before, the suspect changes are:
*  In the new file, there are 12061(!) NUL (^@) characters at the beginning of line 299.
*  Lines 300-322 are exact duplicates of lines 424-446.
*  Lines 323-328 are the most troubling: they are segments of tasks which have been removed from the file elsewhere and mashed together. For example, the line starts with the task which was on line 299, but this task ends midway through its timestamp and another begins.
*  There are a few more lines up to line 456 with similar corruption. e.g. line 456 is just the last 7 characters of a UUID 'xxxxxxx"]'.

Attached are the relevant parts of the last good and current corrupt versions of pending.data (lines 298--396 for the good file, 298--457 for the corrupt file). For privacy, I've replaced alphanumeric characters with 'x' in these files.

I don't know exactly which command caused this to happen. I've got two custom hooks, and I run some commands in an automated or scheduled fashion, but nothing except for taskwarrior touches pending.data directly.

Also, this is the second or third time I've observed similar corruption. Finally, this might be related to TW-253, but I don't use Dropbox.

Cheers!

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2016-01-26T18:35:02Z
Modified: 2017-06-08T12:47:20Z
```

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2016-01-27T01:01:08Z says:_ 

 DropBox does not merge.  Use Taskserver for syncing.

When you stop using hooks and DropBox, do you still experience corruption?

**Comment by taskwarrior** [comment]
_Adam O on 2016-01-27T14:40:00Z says:_ 

 Thanks for getting back to me!

I may not have been clear: I don't use Dropbox. The hooks I have in place don't modify tasks at all: one on-modify hook which prints its second line of input straight out again, and one on-exit hook which copies the contents of .task elsewhere as a backup. I can't say whether the mere existence of these hooks is the problem, but I don't think so; I'm pretty sure the first time I experienced corruption was before I had hooks in place. Certainly, I only set up the on-exit hook after I had already experienced corruption. 

Do you think it's possible that the cause might be multiple concurrent taskwarrior instances?

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2016-01-31T19:54:48Z says:_ 

 Thanks for the info.  I think concurrent taskwarrior instances is the cause, and with the hook scripts, which slow down processing, thereby leaving a larger window of opportunity, I think the file locking is not working here.

**Comment by taskwarrior** [comment]
_Jason Cowley on 2017-06-07T16:30:06Z says:_ 

 I've been getting this problem on my Mac quite frequently using task 2.5.1. It always appears to be the pending.data file that gets corrupted. I'm not using dropbox for sync or any hooks, but I manually sync with inthe.AM fairly frequently. In this case I ran `task next` which worked fine. Then a few minutes later I ran `task 502 done`, followed a few moments later by `task`, after which I got the error.
{code}
...
4 tasks
There are local changes.  Sync required.
~ task 502 done
Completed task 502 'Update readme'.
Completed 1 task.
You have more urgent tasks.
The project 'bridge' has changed.  Project 'bridge' is 50% complete (1 of 2 tasks remaining).
~ task
[task next]
Unrecognized Taskwarrior file format or blank line in data. in /Users/jason/.task/pending.data at line 1
~ task --version
2.5.1
{code}
